src/mlpack/methods/ann/layer/maxout_impl.hpp

```diff
@@ -0,0 +1,137 @@
+/**
+ * @file maxout_impl.hpp
+ * @author Prasanna Patil
+ *
+ * Definition of Maxout layer first introduced in the Maxout Networks,
+ * Ian J. Goodfellow, David Warde-Farley, Mehdi Mirza, Aaron Courville,
+ * Yoshua Bengio
+ * "Maxout Networks", 2013
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "maxout.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout(
+    const size_t inSize,
+    const size_t hiddenSize,
+    const size_t outSize) :
+    inSize(inSize),
+    hiddenSize(hiddenSize),
+    outSize(outSize)
+{
+  weights.set_size((inSize * hiddenSize * outSize) + (outSize * hiddenSize), 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void Maxout<InputDataType, OutputDataType>::Reset()
+{
+  weight = arma::mat(weights.memptr(), outSize * hiddenSize, inSize,
+      false, false);
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+      outSize * hiddenSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (hiddenParameter.n_elem == 0) {
+    hiddenParameter = arma::zeros<arma::mat>(outSize * hiddenSize,
+        input.n_cols);
+  }
+
+  if (output.n_elem == 0) {
+    output = arma::zeros<arma::mat>(outSize, input.n_cols);
+  }
+
+  hiddenParameter = (weight * input) + bias;
```

Comment:
Do we have to initialize the hiddenParameter in line 57 if we just assign the parameter here?